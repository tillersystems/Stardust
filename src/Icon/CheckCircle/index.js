import React from 'react';
import PropTypes from 'prop-types';

/**
 * Check Circle Icon comnponent
 *
 * @return {jsx}
 */
const CheckCircle = ({ fill }) => {
  return (
    <>
      <defs>
        <path
          d="M351.5 41.7a21.3 21.3 0 0 1-17.4 39A192 192 0 1 0 448 256v-19.6a21.3 21.3 0 0 1 42.7 0v19.6A234.7 234.7 0 1 1 351.5 41.7zm133 28.7a21.3 21.3 0 0 1 0 30.1L271 314.1a21.3 21.3 0 0 1-30.2 0l-64-64a21.3 21.3 0 1 1 30.2-30.2l48.9 49L454.2 70.3a21.3 21.3 0 0 1 30.2 0z"
          id="circle"
        />
      </defs>
      <g fill="none" fillRule="evenodd">
        <mask id="check" fill="#fff" fillOpacity="0">
          <use xlinkHref="#circle" />
        </mask>
        <use fill={fill} fillRule="nonzero" xlinkHref="#circle" />
        <g mask="url(#check)" fill={fill}>
          <path d="M0 0h512v512H0z" />
        </g>
      </g>
    </>
  );
};

/**
 * PropTypes Validation
 */
const { string } = PropTypes;

CheckCircle.propTypes = {
  /**
   * Fill color
   */
  fill: string,
};

/**
 * Default props.
 */
CheckCircle.defaultProps = {
  fill: 'white',
};

/**
 * Display name
 */
CheckCircle.displayName = 'CheckCircle';

export default CheckCircle;
