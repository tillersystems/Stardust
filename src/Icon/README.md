# Icon

Icons use graphical symbols to represent an object or concept in your UI. They can be used to aid comprehension of core actions in your app, and to provide feedback for user input.
Stardust provides a large number of pre-built Icon components.

### Icon Props

| Name    | Required |       Type       | DefaultValue |                   Description                   |
| ------- | :------: | :--------------: | :----------: | :---------------------------------------------: |
| `name`  |    +     |      string      |      -       |                    Icon name                    |
| `color` |    -     |      string      |    white     |                   Fill color                    |
| `size`  |    -     | [string, number] |     20px     | Available sizes, including custom - e.g. '20px' |
| `title` |    -     |      string      |     null     |                Alternative text                 |

### Usage

```jsx
import { Icon } from '@tillersystems/stardust';

render() {
  return <Icon name="chevron-left" color="hsl(213,17%,20%)" size="small" title="icon title" />
}
```

<!-- STORY -->
