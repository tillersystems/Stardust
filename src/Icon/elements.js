import styled, { css, keyframes } from 'styled-components';

export const spinAnimation = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

export const Container = styled.div`
  display: inline-block;

  border-image-source: none;
  border-image-width: 0;

  padding: 0;

  font-size: ${({ width }) => width};
  line-height: ${({ width }) => width};

  ${({ spin }) =>
    spin &&
    css`
      animation: ${spinAnimation} 3s linear infinite;
    `};

  ${({ marginRight }) =>
    marginRight &&
    css`
      margin-right: 1rem;
    `};

  ${({ marginLeft }) =>
    marginLeft &&
    css`
      margin-left: 1rem;
    `};
`;
