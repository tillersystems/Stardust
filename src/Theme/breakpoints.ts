/**
 * Defines application breakpoints.
 */

const xs = 0;
const sm = 576;
const md = 768;
const lg = 992;
const xl = 1200;

export const breakpoints = {
  xs,
  sm,
  md,
  lg,
  xl,
};
