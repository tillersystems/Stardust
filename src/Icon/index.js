import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import cuid from 'cuid';

import Theme from '../Theme';
import { Data } from './data';

const { palette } = Theme;

/**
 * Predefined sizes
 */
const SIZE = {
  small: 10,
  medium: 20,
  large: 30,
};

/**
 * Icon comnponent
 *
 * See README, run storybook or see propTypes below for documentation
 *
 * @return {jsx}
 */
const Icon = ({ className, color, name, size, title, ...restProps }) => {
  const generatedId = useMemo(() => cuid(), [title]);
  const titleElementId = `icon-title-${generatedId}`;

  /**
   * Get icon size
   *
   * @return {string}
   */
  const getIconSize = useMemo(() => {
    if (SIZE.hasOwnProperty(size)) return SIZE[size];
    return size;
  }, [size]);

  /**
   * Get color
   *
   * @return {string}
   */
  const getColor = useMemo(() => {
    if (palette.hasOwnProperty(color)) return palette[color];
    return color;
  }, [color]);

  /**
   * Get icon path
   *
   * @return {string}
   */
  const getIconPath = Data[name];

  /**
   * Icon element
   */
  const IconElement = typeof getIconPath === 'function' && getIconPath;

  const rootProps = {
    'aria-hidden': title ? undefined : true,
    'aria-labelledby': title && titleElementId,
    focusable: 'false',
    height: getIconSize,
    role: 'img',
    viewBox: '0 0 512 512',
    width: getIconSize,
    ...restProps,
  };

  return (
    <svg className={className} xmlns="http://www.w3.org/2000/svg" {...rootProps}>
      {title && <title id={titleElementId}>{title}</title>}
      <g>
        {!IconElement ? (
          <path data-testid="icon-svg-path" fill={getColor} fillRule="evenodd" d={getIconPath} />
        ) : (
          <IconElement fill={getColor} />
        )}
      </g>
    </svg>
  );
};

/**
 * PropTypes Validation
 */
const { number, oneOfType, string } = PropTypes;

Icon.propTypes = {
  /**
   * className needed by styled component
   */
  className: string,

  /**
   * Fill color
   */
  color: string,

  /**
   * Icon name
   */
  name: string.isRequired,

  /**
   * Available sizes, including custom - e.g. '20px'
   */
  size: oneOfType([number, string]),

  /**
   * Alternative text
   */
  title: string,
};

/**
 * Default props.
 */
Icon.defaultProps = {
  className: undefined,
  color: 'white',
  size: SIZE.medium,
  title: null,
};

/**
 * Display name
 */
Icon.displayName = 'Icon';

export default styled(Icon)`
  display: inline-block;
`;
