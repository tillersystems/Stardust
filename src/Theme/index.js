/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
