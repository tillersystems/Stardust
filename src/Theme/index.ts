import { DefaultTheme } from 'styled-components';

/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme: DefaultTheme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
