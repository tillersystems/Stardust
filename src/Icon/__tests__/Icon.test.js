import React from 'react';
import 'jest-styled-components';

import Icon from '..';
import Theme from '../../Theme';

describe('<Icon />', () => {
  test('should render without a problem', () => {
    const { container } = render(<Icon name="calendar" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render with different name', () => {
    const { container } = render(<Icon name="check" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render with different size', () => {
    const customSize = '2rem';
    const { getByTestId } = render(<Icon name="calendar" width={customSize} height={customSize} />);
    const iconSvgNode = getByTestId('iconSvg');
    const iconNode = getByTestId('iconContainer');

    expect(iconSvgNode).toHaveAttribute('width', customSize);
    expect(iconSvgNode).toHaveAttribute('height', customSize);
    expect(iconNode).toHaveStyleRule('font-size', customSize);
    expect(iconNode).toHaveStyleRule('line-height', customSize);
  });

  test('should render with predifined color', () => {
    const { getByTestId } = render(<Icon name="calendar" color="darkBlue" />);
    const iconSvgPathNode = getByTestId('iconSvgPath');

    expect(iconSvgPathNode).toHaveAttribute('fill', Theme.palette.darkBlue);
  });

  test('should render with different color', () => {
    const customColor = 'hsl(120, 80%, 20%)';
    const { getByTestId } = render(<Icon name="calendar" color={customColor} />);
    const iconSvgPathNode = getByTestId('iconSvgPath');

    expect(iconSvgPathNode).toHaveAttribute('fill', customColor);
  });

  test('should render with a spin loader', () => {
    const { container } = render(<Icon name="calendar" spin />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
