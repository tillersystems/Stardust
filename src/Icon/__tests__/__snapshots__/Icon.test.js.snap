// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should render with a spin loader 1`] = `
.c0 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
  -webkit-animation: iVXCSc 3s linear infinite;
  animation: iVXCSc 3s linear infinite;
}

<div
  class="c0"
  data-testid="iconContainer"
  height="20"
  width="20"
>
  <svg
    data-testid="iconSvg"
    height="20"
    style="display: block;"
    viewBox="0 0 512 512"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M426.667 192v-64c0-11.782-9.552-21.333-21.334-21.333h-42.666V128c0 11.782-9.552 21.333-21.334 21.333S320 139.783 320 128v-21.333H192V128c0 11.782-9.551 21.333-21.333 21.333-11.782 0-21.334-9.55-21.334-21.333v-21.333h-42.666c-11.782 0-21.334 9.55-21.334 21.333v64h341.334zm0 42.667H85.333v192c0 11.782 9.552 21.333 21.334 21.333h298.666c11.782 0 21.334-9.551 21.334-21.333v-192zM192 64h128V42.667c0-11.782 9.551-21.334 21.333-21.334 11.782 0 21.334 9.552 21.334 21.334V64h42.666c35.347 0 64 28.654 64 64v298.667c0 35.346-28.653 64-64 64H106.667c-35.347 0-64-28.654-64-64V128c0-35.346 28.653-64 64-64h42.666V42.667c0-11.782 9.552-21.334 21.334-21.334S192 30.885 192 42.667V64z"
      data-testid="iconSvgPath"
      fill="hsl(0,100%,100%)"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;

exports[`<Icon /> should render with different name 1`] = `
.c0 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

<div
  class="c0"
  data-testid="iconContainer"
  height="20"
  width="20"
>
  <svg
    data-testid="iconSvg"
    height="20"
    style="display: block;"
    viewBox="0 0 512 512"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
      data-testid="iconSvgPath"
      fill="hsl(0,100%,100%)"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;

exports[`<Icon /> should render without a problem 1`] = `
.c0 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

<div
  class="c0"
  data-testid="iconContainer"
  height="20"
  width="20"
>
  <svg
    data-testid="iconSvg"
    height="20"
    style="display: block;"
    viewBox="0 0 512 512"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M426.667 192v-64c0-11.782-9.552-21.333-21.334-21.333h-42.666V128c0 11.782-9.552 21.333-21.334 21.333S320 139.783 320 128v-21.333H192V128c0 11.782-9.551 21.333-21.333 21.333-11.782 0-21.334-9.55-21.334-21.333v-21.333h-42.666c-11.782 0-21.334 9.55-21.334 21.333v64h341.334zm0 42.667H85.333v192c0 11.782 9.552 21.333 21.334 21.333h298.666c11.782 0 21.334-9.551 21.334-21.333v-192zM192 64h128V42.667c0-11.782 9.551-21.334 21.333-21.334 11.782 0 21.334 9.552 21.334 21.334V64h42.666c35.347 0 64 28.654 64 64v298.667c0 35.346-28.653 64-64 64H106.667c-35.347 0-64-28.654-64-64V128c0-35.346 28.653-64 64-64h42.666V42.667c0-11.782 9.552-21.334 21.334-21.334S192 30.885 192 42.667V64z"
      data-testid="iconSvgPath"
      fill="hsl(0,100%,100%)"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;
